src/testing/run.ts

```diff
@@ -0,0 +1,268 @@
+  }
+}
+
+async function evaluateOutput<TestCaseType, OutputType>(args: {
+  testId: string;
+  testCase: TestCaseType;
+  testCaseHash: string;
+  output: OutputType;
+  evaluator: BaseTestEvaluator<TestCaseType, OutputType>;
+}): Promise<void> {
+  try {
+    const evaluation = await args.evaluator.evaluateTestCase({
+      testCase: args.testCase,
+      output: args.output,
+    });
+    await client.post('/evals', {
+      testExternalId: args.testId,
+      testCaseHash: args.testCaseHash,
+      evaluatorExternalId: args.evaluator.id,
+      score: evaluation.score,
+      threshold: evaluation.threshold,
+    });
+  } catch (err) {
+    await sendError({
+      testId: args.testId,
+      testCaseHash: args.testCaseHash,
+      evaluatorId: args.evaluator.id,
+      error: err,
+    });
+  }
+}
+
+async function runTestCase<TestCaseType, OutputType>(args: {
+  testId: string;
+  testCase: TestCaseType;
+  testCaseHash: string;
+  evaluators: BaseTestEvaluator<TestCaseType, OutputType>[];
+  fn: (testCase: TestCaseType) => OutputType | Promise<OutputType>;
+  maxEvaluatorConcurrency: number;
+}): Promise<void> {
+  let output: OutputType | undefined = undefined;
+
+  try {
+    output = await args.fn(args.testCase);
+  } catch (err) {
+    await sendError({
+      testId: args.testId,
+      testCaseHash: args.testCaseHash,
+      evaluatorId: null,
+      error: err,
+    });
+  }
+
+  if (output === undefined) return;
+
+  await client.post('/results', {
+    testExternalId: args.testId,
+    testCaseHash: args.testCaseHash,
+    testCaseBody: args.testCase,
+    testCaseOutput: isPrimitive(output) ? output : JSON.stringify(output),
+  });
+
+  try {
+    await gatherWithMaxConcurrency({
+      maxConcurrency: args.maxEvaluatorConcurrency,
+      tasks: args.evaluators.map(
+        (evaluator) => () =>
+          evaluateOutput({
+            testId: args.testId,
+            testCase: args.testCase,
+            testCaseHash: args.testCaseHash,
+            output,
+            evaluator,
+          }),
+      ),
+    });
+  } catch (err) {
+    await sendError({
+      testId: args.testId,
+      testCaseHash: args.testCaseHash,
+      evaluatorId: null,
+      error: err,
+    });
+  }
+}
+
+export async function runTestSuite<
+  // A workaround for making the type parameters required
+  // See https://stackoverflow.com/a/76821931
+  T = 'A test case type is required.',
+  O = 'An output type is required.',
+  TestCaseType extends T = T,
+  OutputType extends O = O,
+>(args: {
+  id: string;
+  testCases: TestCaseType[];
+  testCaseHash: // This type requires that the test case hash array has at least one element
+  | [keyof TestCaseType & string, ...(keyof TestCaseType & string)[]]
+    // Or, the user can define their own function to compute the hash
+    | ((testCase: TestCaseType) => string);
+  evaluators: BaseTestEvaluator<TestCaseType, OutputType>[];
+  fn: (testCase: TestCaseType) => OutputType | Promise<OutputType>;
```

Comment:
do you think we'd also want to make `fn` accept an object like `fn({ testCase })` in case we want to add optional args sometime down the road? @dmorton2297 @adamnolte

Reply:
ya I think that's a good idea and would follow the same pattern in evaluators